### ISSUE TITLE
gradient clip for optimizer

### ISSUE THREAD

**Issue by glample** [pytorch/pytorch#309]
We discussed this already, but could we reconsider adding gradient clipping (for L2 and L_inf norms) to pytorch optimizer? This is a very important parameter.. I encountered many problems (in both NLP and especially RL) where the model would totally fail to converge without gradient clipping, independently of the optimizer used.

It's possible to do it now using closure, or even manually before calling optimizer.step(), but it's really not convenient.

**Comment by windweller** [comment]
Hi,

I'm still slightly confused on how to use this API properly and my code is as follows:

```python
optimizer.zero_grad()        
loss, hidden = model(data, hidden, targets)
loss.backward()

torch.nn.utils.clip_grad_norm(model.parameters(), args.clip)
optimizer.step()
```

Does this perform gradient clipping?

**Comment by fmassa** [comment]
@windweller yes, it will perform gradient clipping, see the code implementation in https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/clip_grad.py
We use the github issues for bugs and feature requests, for questions it's better to ask them on https://discuss.pytorch.org/

**Comment by windweller** [comment]
Noted. Thank you :)

**Comment by BoPang1996** [comment]
what is the proper clip for clip_grad_norm? i am new to this. @ @

**Comment by ezelikman** [comment]
For those reading this thread, note that clip_grad_norm is now deprecated for clip_grad_norm_
